### DISCUSSION TITLE
React-flow custom (any) element inside custom node not re-rendering (exported into python Dash)

### DISCUSSION THREAD

**Discussion by rojikada** [xyflow/xyflow#2818] [General]
Hello, I already asked this question on stackoverflow, however it seems like the community is not there yet [(stackoverflow)](https://stackoverflow.com/questions/75337874/plotly-dash-app-callback-does-not-re-render-dash-elements-inside-custom-compo).

I run into problem when I want to have a graph from Dash inside custom anynode:
```
function AnyContentNode({ data }) {
  return (
    <div className="any-content-node" id={data.id} >
      <div>
          {
              data.elements?.map((node, index) => {
                  return window.elements[data.elementId][node];
              })
          }
      </div>
        {
            data.handles?.map((node, index) => {
              return <Handle key={index} type={node.direction} position={node.position} id={node.id} />
            })
        }
    </div>
  );
}
```

I pass the elements to reactflow and store them inside window.elements[some_unique_id] as I found no other way to pass elements to nodes, as data can contain only basic types.

However, when added real-time updates to the custom graph (1s/update), the graph does not re-render unless I click the node and then another node and back to this node.
The componentDidUpdate gets fired properly (many times at once each 1s/whatever time I set it to), but it does not re-render the custom element. Would you happen to know, how to trigger an re-render update please/or is there a better way to propagate elements to a node?

Thanks!
As a side note, I don't have much expertise in React, at first I just wanted to create a (custom wrapper for Python, with the addition of custom node).

**Comment by bcakmakoglu** [comment]
It's generally not that good of an idea to store data inside the window object anyway.

An option would be to use a context and wrap your graph with it: https://reactjs.org/docs/context.html

Then you could use the context to access whatever state you want to have shared across your nodes and access the data from there.
